File: src/main/java/com/vc/medium/L338.scala
package com.vc.medium

object L338 {
  def countBits(num: Int): Array[Int] = {
    val arr = new Array[Int](num + 1)
    (1 to num).foreach(i => {
      arr(i) = arr(i >> 1) + (i & 1) //i >> 1 means divide by two and (i & 1) means add 1 to answer if number is even
    })
    arr
  }
}

File: src/main/java/com/vc/medium/L338.java
package com.vc.medium;

class L338 {
    public int[] countBits(int num) {
        /**
            x = x / 2 + x % 2;
         */
        int[] res = new int[num + 1];
        res[0] = 0;
        for(int i = 1; i <= num; i++) {
            res[i] = res[i >> 1] + (i & 1);
        }
        return res;
    }
}
